vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java

```diff
@@ -283,7 +283,6 @@ public void expanded_nodes_stay_expanded_when_parent_expand_state_is_toggled() {
         Assert.assertEquals(9, getTreeGrid().getRowCount());
 
         getTreeGrid().expandWithClick(7);
-        getTreeGrid().expandWithClick(8);
```

Comment:
Does this mean that an existing test breaks with the change in place?

There should also be a new test that 1. fails without the change 2. passes with the change

Reply:
My understanding is that existing test is faulty and this line should not be there. The bug I am fixing has allowed the existing test to pass. I have asked internally confirmation about this from the original author of the test.

Reply:
> There should also be a new test that 1. fails without the change 2. passes with the change

Done

Reply:
After the change, the Grid on the test page behaves quite unexpectedly in general. There may be applications out there that also rely on the current item identification logic (based on hashCode) and after we'd release this change, they could also start acting weird.

After the change:

https://user-images.githubusercontent.com/1222264/150156899-c0737864-1a7f-4a13-b793-2fa076e57026.mp4

Before the change:

https://user-images.githubusercontent.com/1222264/150156919-a3b2cbb4-60c8-4c69-a274-3f682980f055.mp4

Also, if we change the item ID logic in `TreeGrid`, shouldn't we also change it in `Grid` (and any other component that uses items) for consistency?

Reply:
Thanks for this, it gave me an idea to check one thing.

Reply:
I think the problem is in test page

https://github.com/vaadin/flow-components/blob/fix2383/vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java#L170

There is a bit rigged identity logic in the test entity, and there are more than one bean with id == null. And that is actually causing the problem. Thus the test bean is not realistic and this should be fixed.

Reply:
Are you still working on this? The fix to `TreeGrid` lgtm but this change to the test case isn't correct. As you suggested, the test fix should be done to the `HierarchicalTestBean` identity logic.

Reply:
Lets see, if it works now